### PR TITLE
Building

### PR DESCRIPTION
**Pull request by sondrele**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -22,4 +22,7 @@
 *.fls
 *.out.ps
 *.pyg
-.standalone.tex
```
Comment (havardh) [inline .gitignore]: I still need this one.

File: implementation/building.tex [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+These are, similarly with the EFM32 series, also based on the Cortex-M series of ARM processor cores.
+
+\todo{Håvard: Skriv denne seksjonen}
+The project got up and running pretty fast based on how \texttt{armboot} was built.
+We looked at armboot's Makefile to figure out what flags to pass to \texttt{rustc} in order to make it cross-compile for an ARM target architecture.
+
+\subsection{Transitioning to Cargo}
+\label{ssub:transitioning_to_cargo}
+
+It was always a goal to use Cargo for building, distributing and managing the packages and dependencies that would become part of this project.
```
Comment (havardh) [inline implementation/building.tex]: Include a why for this: 
- To lower the bar for a Rust programmer to make use of it.
- To see/evaluate if it was possible

File: implementation/building.tex [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+While this approach worked, it was far from optimal for a number of reasons:
+
+\begin{itemize}
+    \item Rust was in active development and many of it's unstable \glspl{api} were going through rapid changes. Ensuring that versions of rustc and the Rust source code stayed up to date across different systems was not easy.
+    \item Compiling and ensuring that all dependencies were consistent across builds and systems for the bindings were a tedious task. A lot of the troubles concerning this came back to the point above.
+    \item Linking dependencies with the library required each system to have set up several different \$PATHS to point to the right directories, what worked for one developer on one system might not have worked for different developer on another system.
+    \item The Cargo package manager was developed for exactly these purposes among others.
+\end{itemize}
+
+As already described, Cargo is a tool that provides many operations to build Rust projects that has a certain project structure.
```
Comment (havardh) [inline implementation/building.tex:38]: Reference to section `\autoref{ssub:project_structure}`

File: implementation/building.tex [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+It is designed to integrate with other existing tools, like GNU Make, which has been important in  building this project.
+When the transition to Cargo started, we focused on structuring the main library and its modules into the directory structure described in \autoref{ssub:project_structure}.
+By invoking the \texttt{cargo build --verbose} command, it was possible to see the output from what Cargo attempted to build when it failed, and then structure the project accordingly.
+
+A big priority was to to shrink the size of the makefiles that was in the project by making them a part of the standard build process for \texttt{emlib} instead, doing this would help us get a long way of ensuring that the builds done by Cargo could be consistent across systems.
+By defining a Rust build script and utilizing a Rust build-dependency called \texttt{gcc} \cite{web:cargo_gcc} (not to be mistaken with the C-compiler with the same name), we were able to compile the C sources from Silabs' emlib and link them with our bindings directly as part of the build process.
```
Comment (havardh) [inline implementation/building.tex]: Isn't `gcc-rs` a shell binding for gcc ? the parenthesis here makes it sound like it a whole separate thing.

File: implementation/building.tex [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+By removing the dependency of manually compiling the C sources, it was easier to start to automatically fetch the other dependencies like the \texttt{core} and \texttt{collections} libraries.
+
+Because the project is for a different processor architecture than the system that it is built on, we had to conditionally cross-compile all the standard Rust libraries that we wanted to utilize for the ARM Cortex-M3.
+We could not utilize the pre-compiled libraries that are already included with \texttt{rustc}, since these only works for the current system architecture.
+We solved this problem by implementing a new Cargo build-dependency, called \texttt{rust-src} \cite{github:rust_src}, whose purpose is to download the entire Rust source code that is compliant with the instance of \texttt{rustc} that is currently compiling the library.
```
Comment (havardh) [inline implementation/building.tex]: Do you have any thoughts on improving this process in Rust ? Perhaps something to discuss? How is the sources download at the moment? Does it download only the sub folder for e.g. libcore or does it include the whole source for each library? In that case its quite inefficient.

File: implementation/building.tex [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+By making it a task for each build to fetch its own source code, we were guaranteed that the dependencies we used for the project would always compile, independent of the current instance of \texttt{rustc} that was installed on the system.
```
Comment (havardh) [inline implementation/building.tex:52]: Ah, now I got the point you are making. About having to have a local clone of the rust repo that is in sync with the current compiler. This has to be made more explicit.

File: implementation/building.tex [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+\autoref{tab:compiled_libraries} summarizes the different crates that we have fetched from Rust's standard library in order to use them with our project.
+\todo{May not need this table here. Need to synchronize with the section that presents REL and such..}
+
+\todo{We need to write about the \texttt{rlibc} crate, and the need to define our own \texttt{memmove} and \texttt{memalign} functions in order to use alloc (Box, Vector, ...)}
```
Comment (havardh) [inline implementation/building.tex:56]: The `rlibc` crate might be possible to remove as `newlib` provies `memset` and `memmove`. `rlibc` does in fact not contain `memalign`.

File: implementation/building.tex [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+\end{tabular}
+\caption{Rust libraries conditionally compiled for the Cortex-M3 architecture.}
+\label{tab:compiled_libraries}
+\end{center}
+\end{table}
+
+By design Cargo only supported passing two flags further on to \texttt{rustc}, those were \texttt{-L} and \texttt{-l}, their purpose is to tell \texttt{rustc} to link with an external library by looking in a directory (specified with the \texttt{-L} flag), for a library with the specified name (specified with the \texttt{-l} flag).
+The last step in the build process involved linking \texttt{emlib} and the other libraries with an actual executable for the Cortex-M3.
+This was not possible to do with Cargo since it required us to pass a couple of extra linker-flags further on to \texttt{rustc}.
+\todo{Håvard: Jeg har ingen anelse over hvor riktig dette er. Noe av dette blir vel beskrevet et annet sted, vi får finne en måte å referere til riktig seksjon her, uten å gå altfor mye i detalj...}
```
Comment (havardh) [inline implementation/building.tex]: Dette er en implementasjons detail synes jeg og skal dærfor være her.

File: implementation/building.tex [hunk 9]
```diff
@@ -0,0 +1,165 @@
+\section{Build System}
+\label{sec:build_system}
+
+Building the library and the EFM32 executables turned out to be one of the major parts of the work done in this project.
+This section describes how building the project have changed over time.
+This includes managing dependencies, compiling the RCL and REL and the Silicon Labs EFM32 \texttt{emlib}.
+We have also utilized a continuous integration system that has helped us to keep the project up to date with the nightly builds of Rust and making sure that the builds have been consistent across the systems it has been built on.
+
```
Comment (havardh) [inline implementation/building.tex:8]: Include a birds eye view of the migration process to tie the section together

Ex: `Makefile (rustc+gcc) =>  ... => cargo+linkargs => cargo rustc`

File: implementation/building.tex [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+We worked around this problem by modifying the build script to store the hash generated by Cargo for \texttt{emlib} to a text file, every time the library was built, and then included it in the makefile for the project.
+
+\subsection{Conditional linking with Cargo}
+
+The build process described in the previous section made it simpler to use third party libraries, but it did not solve all of our issues, the main problem that persisted was to have a good way of making \texttt{emlib} itself, portable.
+With the setup that we had, it was easy to create new executables \emph{within} the project, but it was hard to create new executables that \emph{depended} on \texttt{emlib}.
+Basically, because we had to work around Cargo in the final part of the build process, it also meant that \emph{every} project that wanted to depend on \texttt{emlib} also had implement the same workarounds.
+Thus, we needed to solve the problem of knowing where Cargo would store the project metadata, and we needed a way to get Cargo to compile the final executables with the extra linker-arguments needed by \texttt{rustc} in order to compile the binary for the Cortex-M3.
```
Comment (havardh) [inline implementation/building.tex:93]: I like this interlude :)

File: implementation/building.tex [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+
+Cargo does not have much documentation over how its internal works, or how to interfere with the build process, but the documentation does mention that if Cargo is to be invoked with a command that it does not have by default, it will query the system for this command.
+This means that if Cargo is invoked with e.g. the command \texttt{cargo foo <args>...}, it will query the system for an executable with the name \texttt{cargo-foo} and it will invoke this command with the trailing arguments if it exists.
+By looking at Cargo's source code, we could see that every triggered build included a structure called \texttt{CompileOptions}.
+The arguments passed to Cargo's different build commands are then used to compose this structure and trigger an internal compilation process, this process handles the compilation of all dependencies and generates all the different targets for the current package to be compiled.
+\todo{Would it be nice with some kind of figure that describes the build process?}
```
Comment (havardh) [inline implementation/building.tex:100]: +1

File: implementation/building.tex [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+With the setup that we had, it was easy to create new executables \emph{within} the project, but it was hard to create new executables that \emph{depended} on \texttt{emlib}.
+Basically, because we had to work around Cargo in the final part of the build process, it also meant that \emph{every} project that wanted to depend on \texttt{emlib} also had implement the same workarounds.
+Thus, we needed to solve the problem of knowing where Cargo would store the project metadata, and we needed a way to get Cargo to compile the final executables with the extra linker-arguments needed by \texttt{rustc} in order to compile the binary for the Cortex-M3.
+
+Cargo does not have much documentation over how its internal works, or how to interfere with the build process, but the documentation does mention that if Cargo is to be invoked with a command that it does not have by default, it will query the system for this command.
+This means that if Cargo is invoked with e.g. the command \texttt{cargo foo <args>...}, it will query the system for an executable with the name \texttt{cargo-foo} and it will invoke this command with the trailing arguments if it exists.
```
Comment (havardh) [inline implementation/building.tex:97]: Is this just an arbitrary executable? In that case lets mention that this is the conventional way of exposing plugins for cargo.

File: implementation/building.tex [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+\end{tabular}
+\caption{}
+\label{tab:cargo_linkargs}
+\end{center}
+\end{table}
+
+
+In order to solve the problems we had with building the project, we created a new subcommand called \texttt{cargo-linkargs} \cite{github:cargo_linkargs} that depends on Cargo itself.
+This subcommand was created specifically with \texttt{emlib} in mind, and supports all the flags that the \texttt{cargo-build} command supports, including the flags shown in \autoref{tab:cargo_linkargs}.
+We got rid of the two problems we had with building \texttt{emlib} once \texttt{cargo-linkargs} were working.
```
Comment (havardh) [inline implementation/building.tex]: `was working`

File: implementation/building.tex [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+The problem with resolving the location of generated metadata were solved implicitly just by utilizing Cargo, and the extra linker-arguments could easily be passed on to the invocation of \texttt{cargo-linkargs} via the project's makefile.
+
+\subsection{Continuous Integration}
+\label{ssub:continuous_integration}
+
+When we first started this project, Rust had reached a 1.0-alpha version.
```
Comment (havardh) [inline implementation/building.tex:137]: Should we include a Rust road map somewhere?

File: implementation/building.tex [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+This meant that the programming language had reached a relatively stable state, but there was still big parts of the language and its standard libraries that was marked as unstable and up for review before the planned 1.0 release.
+The standard libraries, and third party Rust libraries that has evolved in the Rust community, have made little guarantee about their stability, and the \glspl{api} have been subject to change without much notice.
+
+\todo{Is this section relevant? Too much information? Move elsewhere?}
+
+Many Rust projects have utilized a continuous integration system called Travis CI for ensuring that the code in the project have been compatible with the nightly builds of Rust.
```
Comment (havardh) [inline implementation/building.tex]: Reference to Travis.

File: implementation/building.tex [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+\end{center}
+\end{table}
+
+
+In order to solve the problems we had with building the project, we created a new subcommand called \texttt{cargo-linkargs} \cite{github:cargo_linkargs} that depends on Cargo itself.
+This subcommand was created specifically with \texttt{emlib} in mind, and supports all the flags that the \texttt{cargo-build} command supports, including the flags shown in \autoref{tab:cargo_linkargs}.
+We got rid of the two problems we had with building \texttt{emlib} once \texttt{cargo-linkargs} were working.
+The problem with resolving the location of generated metadata were solved implicitly just by utilizing Cargo, and the extra linker-arguments could easily be passed on to the invocation of \texttt{cargo-linkargs} via the project's makefile.
+
+\subsection{Continuous Integration}
```
Comment (havardh) [inline implementation/building.tex:134]: Add description of CI in general with references

File: implementation/building.tex [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+\label{ssub:continuous_integration}
+
+When we first started this project, Rust had reached a 1.0-alpha version.
+This meant that the programming language had reached a relatively stable state, but there was still big parts of the language and its standard libraries that was marked as unstable and up for review before the planned 1.0 release.
+The standard libraries, and third party Rust libraries that has evolved in the Rust community, have made little guarantee about their stability, and the \glspl{api} have been subject to change without much notice.
+
+\todo{Is this section relevant? Too much information? Move elsewhere?}
+
+Many Rust projects have utilized a continuous integration system called Travis CI for ensuring that the code in the project have been compatible with the nightly builds of Rust.
+By registering our projects with Travis CI, and a community-developed service called Rust CI, we had automatic, daily builds of our projects on a third-party server.
```
Comment (havardh) [inline implementation/building.tex]: Reference to Rust CI

File: implementation/building.tex [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+The problem with resolving the location of generated metadata were solved implicitly just by utilizing Cargo, and the extra linker-arguments could easily be passed on to the invocation of \texttt{cargo-linkargs} via the project's makefile.
+
+\subsection{Continuous Integration}
+\label{ssub:continuous_integration}
+
+When we first started this project, Rust had reached a 1.0-alpha version.
+This meant that the programming language had reached a relatively stable state, but there was still big parts of the language and its standard libraries that was marked as unstable and up for review before the planned 1.0 release.
+The standard libraries, and third party Rust libraries that has evolved in the Rust community, have made little guarantee about their stability, and the \glspl{api} have been subject to change without much notice.
+
+\todo{Is this section relevant? Too much information? Move elsewhere?}
```
Comment (havardh) [inline implementation/building.tex]: It is relevant but should be moved into testing section IMO.

File: implementation/building.tex [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+By making continuous integration part of the normal build routine and review-process for new project code, we had an extra step of verification that the project would build on other systems then the one it was developed on.
+
+It is important to note that continuous integration only helped us to verify that the project could be \emph{built}, it could not help us to verify that the compiled code would actually \emph{work} for its target architecture.
+In order to verify that the code would work for on the Cortex-M3 we had to actually run in on one of the microcontrollers that we had a available for this project.
+An experimental process of testing and mocking the \texttt{emlib} \gls{api} bindings is described in greater detail in \autoref{ssub:testing}.
+
+\subsection{Contributing to Cargo}
+\label{ssub:contributing_to_cargo}
+
+As already mentioned, the ability to pass arbitrary flags down to the invocation of \texttt{rustc} was by design not supported by Cargo, but many people in the Rust community have wanted the ability to do so.
```
Comment (havardh) [inline implementation/building.tex]: Refere to github issues ?

File: implementation/building.tex [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+The reasoning for not allowing arbitrary flags to be passed down is pretty obvious.
+A compilation can go awry very quickly if it is up to the package \emph{author} what flags should be passed to \texttt{rustc} instead of the package \emph{builder}.
```
Comment (havardh) [inline implementation/building.tex]: explain who the "package builder" is

File: implementation/building.tex [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+This would give the author the ability to set the restriction for a library, not the user.
+Different systems do not necessarily support all flags and possibilities, so if a package dependency says that it is to be built in a specific way it might not work for the system that is building it.
+\todo{Should we include references to the issues?}
```
Comment (havardh) [inline implementation/building.tex]: yop.

File: implementation/building.tex [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+
+There have been an issue on Cargo's project page about implementing a new subcommand (called \texttt{cargo-rustc}) for the package manager.
+This subcommand would allowed for passing arbitrary flags further on to \texttt{rustc}, but with the restriction of only compiling a \emph{single} target at a time.
```
Comment (havardh) [inline implementation/building.tex]: what is a target

File: implementation/building.tex [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+This means that only \emph{either} the library, a binary, an example or a test (or a package dependency), may be compiled with the extra flags, and \emph{not} the entire package.
+These rules are restrictive enough to get libraries to not depend on a set of extra flags, but loose enough so that specialized projects, like our \texttt{emlib}, can depend on it for completing the build.
```
Comment (havardh) [inline implementation/building.tex:164]: I like this conclusion.

File: implementation/building.tex [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+Indeed, the functionality proposed with this subcommand was exactly the same as was solved with our implementation of \texttt{cargo-linkargs}.
```
Comment (havardh) [inline implementation/building.tex]: exactly ?

File: implementation/building.tex [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+
+After gaining insight into Cargo's internals during the development of \texttt{cargo-linkargs}, it was interesting to see if we could get this same functionality into Cargo itself by implementing \texttt{cargo-rustc}.
+Not only would it benefit our project, it would also give many other Rust projects the ability to use Cargo for the entire compilation process.
+Since both Rust and Cargo are open source projects, it was easy to get in contact with the project maintainers about the issue, and eventually submit a pull-request with the new subcommand.
+After it was reviewed by one of the project maintainers, the pull-request was accepted and merged into Cargo's code base.
```
Comment (havardh) [inline implementation/building.tex]: Add: `and will be part of the Rust 1.0 distribution` ?

File: implementation/building.tex [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+This subcommand would allowed for passing arbitrary flags further on to \texttt{rustc}, but with the restriction of only compiling a \emph{single} target at a time.
+This means that only \emph{either} the library, a binary, an example or a test (or a package dependency), may be compiled with the extra flags, and \emph{not} the entire package.
+These rules are restrictive enough to get libraries to not depend on a set of extra flags, but loose enough so that specialized projects, like our \texttt{emlib}, can depend on it for completing the build.
+Indeed, the functionality proposed with this subcommand was exactly the same as was solved with our implementation of \texttt{cargo-linkargs}.
+
+After gaining insight into Cargo's internals during the development of \texttt{cargo-linkargs}, it was interesting to see if we could get this same functionality into Cargo itself by implementing \texttt{cargo-rustc}.
+Not only would it benefit our project, it would also give many other Rust projects the ability to use Cargo for the entire compilation process.
```
Comment (havardh) [inline implementation/building.tex:169]: Emphasis the issues with depending on a plugin subcommand for cargo => not ergonomic.

File: implementation/building.tex (hunk 24 above)
Comment (sondrele) [inline implementation/building.tex]: (ಠ_ಠ)

File: implementation/building.tex (hunk 23 above)
Comment (sondrele) [inline implementation/building.tex:164]: ヽ༼ຈل͜ຈ༽ﾉ

File: implementation/building.tex (hunk 22 above)
Comment (sondrele) [inline implementation/building.tex]: :smile: :gun:

File: implementation/building.tex (hunk 7 above)
Comment (sondrele) [inline implementation/building.tex:56]: still need to write about the need for memalign

File: implementation/building.tex (hunk 8 above)
Comment (sondrele) [inline implementation/building.tex]: Ok. Da burde utdype mer hva problemet var her

File: implementation/building.tex (hunk 9 above)
Comment (sondrele) [inline implementation/building.tex:8]: tenkte at det hadde vært fint med en figur for dette!

File: implementation/building.tex (hunk 14 above)
Comment (sondrele) [inline implementation/building.tex:137]: I've been thinking on the exact same thing!

File: implementation/building.tex (hunk 3 above)
Comment (sondrele) [inline implementation/building.tex:38]: Denne seksjonen blir referert senere i samme avsnitt

**Pull request merged**